### DISCUSSION TITLE
How to convert verticalWKID WGS84 height (wkid: 4326, verticalWkid: 115700) to DHHN2016 (wkid 4258, verticalWkid: 7837)

### DISCUSSION THREAD

**Discussion by ecr334** [ngageoint/projections-ios#1] [Q&A] [answered]
Hey! Tell me how to make a transformation using this framework convert WGS84 height (wkid: 4326, verticalWkid: 115700) to DHHN2016 (wkid 4258, verticalWkid: 7837)

**Comment by bosborn** [comment, marked as answer]
The easiest way would be to just transform between the EPSG codes.  The elevation does not make a difference and is simply preserved.  Alternatively you can get more complex by specifying the well-known text definitions (compounds in WKT2 with verticals).  I don't think you'll see much difference [between 4326 and 4258](https://gis.stackexchange.com/questions/199087/projection-from-epsg4326-to-epsg4258) though.

```
    PROJProjectionTransform *transform = [PROJProjectionTransform transformFromEpsg:4326 andToEpsg:4258];
    
    CLLocationCoordinate2D coordinate = CLLocationCoordinate2DMake(52.51064626078732, 13.403729270903389);
    CLLocationCoordinate2D transformedCoordinate = [transform transform:coordinate];
    
    PROJLocationCoordinate3D *coordinate3d = [PROJLocationCoordinate3D coordinateWithCoordinate:coordinate andZ:[[NSDecimalNumber alloc] initWithDouble:100.0]];
    PROJLocationCoordinate3D *transformedCoordinate3d = [transform transform3d:coordinate3d];
```

**Reply by ecr334** [reply]
Thanks for the answer, with your framework, you can't transform heights to different spatial references?
```
guard let coordinateFirst = PROJLocationCoordinate3D(coordinate: CLLocationCoordinate2D(latitude: 48.09763488, longitude: 8.58171633), andZ: 698.458) else { return }
  let proj1 = PROJProjectionFactory.projection(withAuthority: PROJ_AUTHORITY_EPSG, andIntCode: 4326, andParams: "+proj=longlat +datum=WGS84 +no_defs")
 let projEVRF : PROJProjection = PROJProjectionFactory.projection(withAuthority: PROJ_AUTHORITY_EPSG, andIntCode: 4314, andParams: "+proj=longlat +ellps=bessel +towgs84=598.1,73.7,418.2,0.202,0.045,-2.455,6.7 +no_defs")
guard let projTRANS = PROJProjectionTransform(from: proj1, andTo: projEVRF) else { return }
print("proj-Z:", projTRANS.transform3d(coordinateFirst).z)
result: projZ: 645.289987946860544
```

So "transformed height" is not correct?

**Reply by bosborn** [reply]
That transformed height should be correct.  The height is passed into PROJ.4 (version 4.9.3) pj_transform method and the result returned.

I was more thinking when you get in to compound WKT2 (well-known text 2) definitions with a geographic/geodetic CRS and a vertical CRS.  The geo CRS is used in the PROJ.4 conversion.

Such as:
```
COMPOUNDCRS["WGS84 Height (EGM08)",
    GEOGCRS["ETRS89",
        DATUM["European_Terrestrial_Reference_System_1989",
            ELLIPSOID["GRS 1980",6378137,298.257222101,
                ID["EPSG",7019]],
            ID["EPSG",6258]],
        PRIMEM["Greenwich",0,
            ID["EPSG",8901]],
        CS[ellipsoidal,2],
        AXIS["Lon",east],
        AXIS["Lat",north],
        UNIT["degree",0.0174532925199433,
            ID["EPSG",9122]],
        ID["EPSG",4258]],
    VERTCRS["DHHN2016 height",
        VDATUM["Deutsches Haupthoehennetz 2016",
            ID["EPSG",1170]],
        CS[vertical,1,
            ID["EPSG",6499]],
        AXIS["Gravity-related height (H)",up],
        LENGTHUNIT["metre",1,
            ID["EPSG",9001]],
        ID["EPSG",7837]]]
```

The vertical CRS in this compound is not currently used in the transformation.

**Reply by bosborn** [reply]
Also, looking into this I had to fix an issue with non projected ETRS89 CRS parsing (EPSG:4258).  So i'll have to release new versions [with this fix](https://github.com/ngageoint/coordinate-reference-systems-ios/commit/cd236269830380a37c9cf5c9ea536334b3ee8bdf) before WKT definition parsing of EPSG:4258 works.

**Reply by ecr334** [reply]
> The vertical CRS in this compound is not currently used in the transformation.

How do I make it count in the transformation? 

I am making an application that will be able to transform elevations from ellipsoidal to geoid, and I can't yet figure out how to use your framework if I create a PROJProjection using WKT (from arcGIS.WKtext) 

```
GEOGCS["GCS_ETRS_1989",DATUM["D_ETRS_1989",SPHEROID["GRS_1980",6378137.0,298.257222101]],PRIMEM["Greenwich",0.0],UNIT["Degree",0. 0174532925199433]],VERTCS["DHHN2016_(height)",VDATUM["Deutsches_Haupthoehennetz_2016"],PARAMETER["Vertical_Shift",0.0],PARAMETER["Direction",1.0],UNIT["Meter",1.0]]
```

I get an error like this:

```
2022-05-03 13:18:38.441661+0300 proj[13557:348324] Ignored end: ",VERTCS[DHHN2016_(height),VDATUM[Deutsches_Haupthoehennetz_2016],PARAMETER[Vertical_Shift,0.0],PARAMETER[Direction,1.0],UNIT[Meter,1.0]]"
```

Question, where did you get the WKT you inserted in the answer ? 
I'm new to all this, and don't quite understand how it works, or where to look for the information.
I have the start values and end values to get, but using arcgis I didn't get the result I wanted

<img width="500" alt="Снимок экрана 2022-05-03 в 13 23 38" src="https://user-images.githubusercontent.com/16240094/166437868-caff7dc3-23e4-4486-99db-79b0c117b7ed.png">

**Reply by bosborn** [reply]
Anything not defined in a single EPSG or set of PROJ4 parameters ([pre version 5](https://github.com/OSGeo/PROJ/releases/tag/5.0.0)) is not supported for elevation.  The recent few years of PROJ overhaul changes with vertical grid shifts, steps, compounds, etc are not currently supported.  As far as projections and software go, these are all still pretty new changes and not how projections have been done before 2018.  Might try and update to newer versions at some point, but not an easy quick change.  Building the C++ library successfully for iOS use and converting to the newer API's would require some effort.  The Java version of this library is also using a C++ port of PROJ.4 [before the version 5+ PROJ changes](https://github.com/locationtech/proj4j/issues/12).

If a compound is specified in a definition, currently the first geo CRS is used in the transformation.  The [CRS library](https://github.com/ngageoint/coordinate-reference-systems-ios) I wrote last year handles parsing CRS WKT 1 and 2 including compounds.  Converting those more complex / compound CRS definitions to functioning PROJ.4 projections is currently the issue.

The Arc WKT listed does not appear to be valid.  It looks like a valid GEOGCS CRS followed by (using basic concatenation) an invalid (maybe Arc specific) vertical CRS that does not follow CRS WKT 1 (VERT_CS...) or 2 (VERTCRS...).

I looked up WKT versions of the EPSG codes and pasted them into a compound.

[WKT1 of EPSG:4258](https://epsg.io/4258)
[WKT2 of EPSG:4258](https://epsg.org/crs/wkt/id/4258)
[WKT2 of EPSG:7837](https://epsg.org/crs/wkt/id/7837)
[WKT1 of EPSG:5783 (replaced by 7837)](https://epsg.io/5783)
[EPSG:7837 w/ no WKT1 listed](https://epsg.io/7837)

WKT CRS Spec: [Geographic information — Well-known text representation of coordinate reference systems](https://docs.opengeospatial.org/is/18-010r7/18-010r7.html)

I also wrote and use these tools that rely on Java versions of theses libraries to help with CRS WKT formatting and projections
* [CRS Pretty](https://github.com/ngageoint/coordinate-reference-systems-java/tree/master/script/pretty)
* [Project](https://github.com/ngageoint/projections-java/tree/master/script/project)

**Reply by bosborn** [reply]
> Also, looking into this I had to fix an issue with non projected ETRS89 CRS parsing (EPSG:4258). So i'll have to release new versions [with this fix](https://github.com/ngageoint/coordinate-reference-systems-ios/commit/cd236269830380a37c9cf5c9ea536334b3ee8bdf) before WKT definition parsing of EPSG:4258 works.

This issue fixed in version [1.0.3](https://github.com/ngageoint/projections-ios/releases/tag/1.0.3) and [crs-ios version 1.0.3.](https://github.com/ngageoint/coordinate-reference-systems-ios/releases/tag/1.0.3)
